qiskit/ignis/experiments/calibration/instruction_data/gate_and_schedules_definition.py

```diff
@@ -55,243 +78,196 @@ def from_backend(cls, backend: BaseBackend) -> 'InstructionsDefinition':
-        params = {}
-        if pulse_envelope.__name__ in ['Drag', 'Gaussian', 'GaussianSquare']:
-            params['amp'] = Parameter(name + '.' + channel.name + '.' + pulse_name + '.amp')
-            params['sigma'] = Parameter(name + '.' + channel.name + '.' + pulse_name + '.sigma')
-
-        if pulse_envelope.__name__ == 'Drag':
-            params['beta'] = Parameter(name + '.' + channel.name + '.' + pulse_name + '.beta')
-
-        if pulse_envelope.__name__ == 'GaussianSquare':
-            params['width'] = Parameter(name + '.' + channel.name + '.' + pulse_name + '.width')
-
-        return params
-
-    def add_composite_instruction(self, name: str, instructions: List[List[Tuple[str, Tuple]]]):
-        """
-        A composite instruction is an instruction that refers to other instructions.
-        It is specified as a list of list where timing barriers are inserted in between
-        lists. All instructions within a sublist are left aligned.
-
-        Args:
-            name: The name of the composite instruction.
-            instructions: List of sub-lists which contain references to
-                instructions as (name, qubits) keys. Here, name is a string and
-                qubits is a tuple. Each instruction is a schedule. All pulses within
-                 a sub-list are left-aligned and form a sub-schedule. Sub-schedules are
-                 inserted one after the other without any time overlap.
-        """
-
-        qubits = set()
-        for inst_list in instructions:
-            for inst_config in inst_list:
-                qubits |= set(inst_config[1])
-
-        self._instructions[(name, tuple(qubits))] = instructions
-
-    def _get_composite_instruction(self, instructions: List[List[Tuple[str, Tuple]]]) -> Schedule:
-        """
-        Recursive function to obtain a schedule.
+        Returns:
+            Generated gate id for this entry.
         """
-
-        schedule = Schedule()
-        for inst_list in instructions:
-            with pulse.build() as sub_schedule:
-                for inst_config in inst_list:
-                    name = inst_config[0]
-                    qubits = inst_config[1]
-                    inst = self._instructions[(name, qubits)]
-
-                    if not isinstance(inst, Schedule):
-                        sched = self._get_composite_instruction(inst)
-                    else:
-                        sched = inst
-
-                    sub_schedule.append(sched, inplace=True)
-
-            schedule.insert(schedule.duration, sub_schedule, inplace=True)
-
-        return schedule
+        if isinstance(qubits, int):
+            qubits = (qubits, )
+
+        gate_id = self._deduplicate_gate_id(gate_name, qubits)
+
+        self._instructions = self._instructions.append(
+            {'gate_name': gate_name,
+             'qubits': qubits,
+             'signature': signature,
+             'gate_id': gate_id,
+             'program': sched_code},
```

Comment:
What prevents the user from inserting invalid strings here? Should we not parse the DSL to ensure it is valid before inserting?

Reply:
Currently we cannot prevent this because now we directly input `sched_code`. If invalid string is provided, the `get_gate_schedule` method will be crashed. I think we need a logic to generate valid schedule code.